src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -256,4 +252,129 @@ public void onFailure(@NonNull Throwable t) {
+        /** Step nodes mapped to the step execution. Entries removed when they are ready to be resumed. */
+        private final Map<FlowNode, StepExecution> nodes = new HashMap<>();
+        /** Step nodes currently being resumed. Removed after resumption completes. */
+        private final Set<FlowNode> processing = new HashSet<>();
+        /** Step nodes mapped to the nearest enclosing step node (no entry if at root). */
+        private final Map<FlowNode, FlowNode> enclosing = new HashMap<>();
+
+        ParallelResumer(Collection<StepExecution> executions, Runnable onCompletion) {
+            this.onCompletion = onCompletion;
+            // First look up positions in the flow graph, so that we can compute dependencies:
+            for (StepExecution se : executions) {
+                try {
+                    FlowNode n = se.getContext().get(FlowNode.class);
+                    if (n != null) {
+                        nodes.put(n, se);
+                    } else {
+                        LOGGER.warning(() -> "Could not find FlowNode for " + se + " so it will not be resumed");
+                    }
+                } catch (IOException | InterruptedException x) {
+                    LOGGER.log(Level.WARNING, "Could not look up FlowNode for " + se + " so it will not be resumed", x);
+                }
+            }
+            for (FlowNode n : nodes.keySet()) {
+                LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+                scanner.setup(n);
+                for (FlowNode parent : scanner) {
+                    if (parent != n && nodes.containsKey(parent)) {
+                        enclosing.put(n, parent);
+                        break;
+                    }
+                }
+            }
+        }
+
+        synchronized void run() {
+            LOGGER.fine(() -> "Checking status with nodes=" + nodes + " enclosing=" + enclosing + " processing=" + processing);
+            if (nodes.isEmpty()) {
+                if (processing.isEmpty()) {
+                    LOGGER.fine("Done");
+                    onCompletion.run();
+                }
+                return;
+            }
+            Map<FlowNode, StepExecution> ready = new HashMap<>();
+            for (Map.Entry<FlowNode, StepExecution> entry : nodes.entrySet()) {
+                FlowNode n = entry.getKey();
+                FlowNode parent = enclosing.get(n);
+                if (parent == null || !nodes.containsKey(parent)) {
+                    ready.put(n, entry.getValue());
+                }
+            }
+            LOGGER.fine(() -> "Ready to resume: " + ready);
+            nodes.keySet().removeAll(ready.keySet());
+            for (Map.Entry<FlowNode, StepExecution> entry : ready.entrySet()) {
+                FlowNode n = entry.getKey();
+                StepExecution exec = entry.getValue();
+                processing.add(n);
+                Timer.get().submit(() -> {
```

Comment:
Probably should not use `Timer` for this, as it is throttled to 10 threads. Could use `threadPoolForRemoting`, or expose `SynchronousNonBlockingStepExecution.executorService` (see https://github.com/jenkinsci/workflow-step-api-plugin/pull/110).